### PR TITLE
Zatrazz/cortex strings import

### PR DESCRIPTION
**Pull request by zatrazz**
Regarding licensing, I added as-in since there are files that we, Linaro, can not re-license (the Copyright is set for ARM). So I think a better strategy would be to evaluate each implementation to see which one ARM will need to change the header license and which one I can re-license myself.

I have not included the SVE implementations.

### CODE REVIEW

**Comment by zatrazz** [comment]
Ping.

**Comment by nsz-arm** [comment]
sorry, i was away, the patch set looks good, but i will have to think about the licensing bit.

File: string/test/memchr.c [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+#include <limits.h>
+#include "stringlib.h"
+
+static const struct fun
+{
+	const char *name;
+	void *(*fun)(const void *, int c, size_t n);
+} funtab[] = {
+#define F(x) {#x, x},
+F(memchr)
```
Comment (nsz-arm) [inline string/test/memchr.c]: this should test __memchr_aarch64 and __memchr_arm

File: string/test/memcmp.c [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+#include <string.h>
+#include "stringlib.h"
+
+static const struct fun
+{
+	const char *name;
+	int (*fun)(const void *s1, const void *s2, size_t n);
+} funtab[] = {
+#define F(x) {#x, x},
+F(memcmp)
```
Comment (nsz-arm) [inline string/test/memcmp.c]: same as memchr

File: string/test/memchr.c (hunk 1 above)
Comment (nsz-arm) [reply string/test/memchr.c]: sorry i see that is added later

**Comment by nsz-arm** [comment]
if you drop the opt_barrier patch then i will merge the rest.

about licenses: please prepare a separate patch that just removes all old copyright headers and replaces them with the new MIT header, as far as i can tell everything is owned by linaro or arm so i can approve such a patch.

**Comment by zatrazz** [comment]
> i don't like this change, it breaks opt_barrier_* on !**GNUC** compilers, i'd just add the "m" variants without changing existing code.

It does not really change anything for !**GNUC** compiler, it will keep using the volatile way.

> 
> you can have a macro for the asm constraint e.g. BARRIER(x) to avoid some repetition but it should just expand to the asm statement.

That is exactly what it does, the opt_barrier/force_eval will expand to the asm statemente for __GNUC__ compiler, or volatile otherwise.

> 
> optimizing away the + is not important, force_eval is used when the arg is otherwise unused so it's unlikely that the compiler cares if it is clobbered or not.
> 
> in any case this is not relevant for the string function import.

Alright, I don't mind drop this change. I add mainly to avoid the above warning on gcc for arm:

math/math_config.h: In function ‘force_eval_float’:
math/math_config.h:188:18: warning: unused variable ‘y’ [-Wunused-variable]
  188 |   volatile float y = x;
      |                  ^

I think it is better than keep adding architecture defines on logic to use the asm statement (assuming other architecture would use this library).

**Comment by zatrazz** [comment]
> if you drop the opt_barrier patch then i will merge the rest.
> 
> about licenses: please prepare a separate patch that just removes all old copyright headers and replaces them with the new MIT header, as far as i can tell everything is owned by linaro or arm so i can approve such a patch.

Alright, I will work on this.

**Comment by maxim-kuvyrkov** [comment]
> if you drop the opt_barrier patch then i will merge the rest.
> 
> about licenses: please prepare a separate patch that just removes all old copyright headers and replaces them with the new MIT header, as far as i can tell everything is owned by linaro or arm so i can approve such a patch.

@nsz-arm , Linaro can change license only on files that Linaro has copyright on.  For the few files that are copyrighted by Arm it needs to be you or someone else from Arm to submit the patch changing license.

@zatrazz , would you please prepare the patch to change license header on all Linaro-copyrighted files?

**Pull request closed without merging**